---
title: "Mean TCRen for TCR repertoires"
author: "Vadim Karnaukhov"
date: "09 06 2021"
output: html_document
---


# Read dataframe with pairwise potential
File contains 3 columns: "residue.aa.from" (amino acid on TCR side), "residue.aa.to" (amino acid on peptide side), "TCRen" (value of pairwise energy of interaction; the lower, the more energetically favorable)

```{r}
potential <- fread("TCRen.csv")
```


# Amino acid composition (AAComp) of proteomes of different species (including human)
You can use either precomputed AAComp of human proteome or calculate AAComp directly from Uniprot proteome fasta file

Precomputed amino acid composition of human proteome
```{r}
AAComp <- tibble(aminoacid = c("A","R","N","D","C","E","Q","G","H","I","L","K","M","F","P","S","T","W","Y","V"),
                 fraction = c(0.070,0.056,0.036,0.047,0.023,0.071,0.048,0.066,0.026,0.043,0.100,0.057,0.021,0.037,0.063,0.083,0.054,0.012,0.027,0.060),
                 organism = "Human")
```

Calculate AAComp from Uniprot proteome fasta file
```{r}
library(protr)
library(magrittr)

#function for calculation of amino acid composition starting from proteome fasta file
compute_AAComp <- function(proteome_fasta_file, organism_name) {
  readFASTA(proteome_fasta_file, seqonly = T) %>% 
    paste0(collapse = "") %>% as.character() %>% 
    gsub(pattern = "[XU]", replacement = "") %>% #delete non-standard amino acids
    extractAAC() %>%
    as.data.frame() %>% 
    rownames_to_column() %>% 
    set_colnames(c("aminoacid", "fraction")) %>% 
    mutate(organism = organism_name)
}

AAComp <- rbind(
  compute_AAComp("proteomes/UP000005640_9606.fasta", "Human"),
  compute_AAComp("proteomes/UP000002241_11706.fasta", "HIV"),
  compute_AAComp("proteomes/UP000000938_295027.fasta", "EBV"),
  compute_AAComp("proteomes/UP000000558_83334.fasta", "EColi")
)
```


# Compute average values of TCRen potential for residues in TCR side
Normalized on amino acid composition in different species

```{r}
potential.mean <- potential %>% 
  merge(AAComp %>% 
          select(residue.aa.to = aminoacid, fraction, organism)) %>% 
  group_by(residue.aa.from, organism) %>% 
  summarise(TCRen.mean = sum(TCRen * fraction))
```


# Calculate TCRen.mean for CDR3's in TCR repertoire

```{r}
# Function for calculation of average TCRen for CDR3's in repertoire
# Input: a dataframe with "freq" and "cdr3aa" columns (e.g. in MiXCR format); name of the organism, we use TCRen.mean values with normalisation on amino acid composition of this organism, default is "Human"
# Output: a dataframe with 4 columns "freq", "cdr3aa" (both from the input file), "TCRen.mean.sum" (sum of TCRen.mean values for CDR3 residues), "TCRen.mean.mean" (average TCRen.mean value for CDR3 residues)
# Consider CDR3 sequence without first and last 4 residues which almost never contact peptide

TCRen.mean_calc <- function(df, organism = "Human") {
  TCRrep <- df %>% 
    select(freq, cdr3aa) %>% 
    mutate(cdr3aa.trim = substr(cdr3aa, 5, nchar(cdr3aa) - 4)) 
  
  max.len.CDR3 <- TCRrep$cdr3aa.trim %>% nchar() %>% max() 
  
  TCRrep %>% 
    separate(cdr3aa.trim, into = paste0("p", 1:max.len.CDR3), 
                      sep = 1:max.len.CDR3, remove = F) %>% 
             melt(id = c("freq", "cdr3aa", "cdr3aa.trim"), variable.name = "position", value.name = "aminoacid") %>% 
    merge(potential.mean %>% 
            filter(organism == organism) %>% 
            select(aminoacid = residue.aa.from, TCRen.mean)) %>% 
    group_by(freq, cdr3aa) %>% 
    summarise(TCRen.mean.sum = sum(TCRen.mean),
              TCRen.mean.mean = mean(TCRen.mean))
}
```

Function calling
```{r}
TCRen.mean_calc(fread("Donor1_1_TRB_Tfh.txt"))
```

# Example of comparision of TCRen.mean of TCR repertoires from different subsets (data from Kasatskaya et al., Elife, 2020)

```{r}
# Download data from https://figshare.com/s/2145b1b16c6854445af7 

# Path to the dir with data from Kasatskaya et al., Elife (2020)
dir_name <- "~/shared-with-me/struct2020/benchmark/subsets"

# Consider only top-N clones by freq
N <- 3000
```

Prepare annotation for files
```{r}
file_list <- dir(dir_name) %>% 
  sapply(function(x) startsWith(prefix = "Donor", x = x) &
           x != "Donor4_3_2_TRB_Th1-17.txt")

file_list <- dir(dir_name)[file_list]

kasatskaya <- tibble(file_name = file_list) %>% 
  mutate(Donor = str_split_fixed(file_name, "_", 5)[,1],
         chain = str_split_fixed(file_name, "_", 5)[,3],
         subset = str_split_fixed(file_name, "_", 5)[,4],
         subset = substr(subset, 1, nchar(subset) - 4)) %>% 
  filter(chain %in% c("TRA", "TRB")) %>% 
  # assign subsets to larger groups
  merge(tibble(subset = c("Tfh","TREG","Th1-17","Th2a","Th17","Th2b","Th22","Th1"),
               subset.group = c("Tfh","TREG","Th1","Th2","Th17","Th2","Th17","Th1")))

```


```{r}
# Compute average TCRen.mean.sum and TCRen.mean.mean for whole repertoires

kasatskaya.TCRen <- kasatskaya$file_name %>% 
  lapply(function(x) fread(paste0("~/shared-with-me/struct2020/benchmark/subsets/", x)) %>% 
           arrange(-freq) %>% 
           head(N) %>% 
           TCRen.mean_calc() %>% 
           mutate(file_name = x)
         ) %>% 
  rbindlist() %>% 
  group_by(file_name) %>% 
  summarise(rep.TCRen.mean.sum = mean(TCRen.mean.sum),
            rep.TCRen.mean.mean = mean(TCRen.mean.mean))
```

Consider larger subsets (Th2 as Th2a+Th2b, Th1 as Th1+Th1-17, Th17 as Th17+Th22)
```{r}
kasatskaya.TCRen.s. <- kasatskaya.TCRen %>% 
  merge(kasatskaya) %>% 
  group_by(subset.group, Donor, chain) %>% 
  summarise(rep.TCRen.mean.sum = mean(rep.TCRen.mean.sum),
            rep.TCRen.mean.mean = mean(rep.TCRen.mean.mean)) 
```

Plot repertoire mean TCRen 
```{r}
kasatskaya.TCRen.s. %>% 
  melt(id = c("subset.group", "Donor", "chain")) %>% 
  ggplot(aes(x = subset.group, y = value)) +
  facet_wrap(chain ~ variable, ncol = 2, scales = "free") +
  geom_boxplot() +
  geom_point() +
  theme_bw()

kasatskaya.TCRen.s. %>% 
  melt(id = c("subset.group", "Donor", "chain")) %>% 
  dcast(subset.group + Donor + variable ~ chain) %>% 
  ggplot(aes(x = TRA, y = TRB, group = subset.group, color = subset.group, fill = subset.group)) +
  geom_point() +
  geom_polygon(alpha = .2) +
  facet_wrap(~variable, scales = "free") +
  theme_bw()
```


# Supplementary plots

Plot differencies in AAComp between different species and human 
```{r}
aa.levels <- c("L","I","M","V","F","W","Y","C","H","A","G","P","T","S","Q","N","D","E","R","K")

AAComp %>% 
  filter(organism != "Human") %>% 
  merge(AAComp %>% 
          filter(organism == "Human") %>% 
          select(aminoacid, fraction.human = fraction)) %>% 
  mutate(dif = fraction - fraction.human) %>% 
  ggplot(aes(x = factor(aminoacid, levels = aa.levels), y = dif, color = organism, group = organism)) +
  geom_line() +
  geom_hline(yintercept = 0) +
  xlab("") + ylab("AAComp(organism) - AAComp(human)") +
  theme_bw()
```

```{r}
#AAComp %>% filter(organism == "Human") %>% select(type = organism, aminoacid, freq = fraction) %>% 
#  rbind(aa.freq.interface) %>% 
#  ggplot(aes(x = factor(aminoacid, levels = aa.levels), y = freq, color = type, group = type)) +
#  geom_line() 
```

Plot TCRen.mean for different species
```{r}
potential.mean %>% 
  ggplot(aes(x = factor(residue.aa.from, levels = aa.levels), y = TCRen.mean, color = organism, group = organism)) +
  geom_line() +
  xlab("") + 
  theme_bw()
```